### PR TITLE
Small bank balance as a Read handler

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves https://github.com/microsoft/CCF/issues/417.

These changes result from switching the `balance` endpoint of our small bank app from `Write` to `Read`:
- The primary (and only the primary) now stores the signed client request.
- For a forwarded command, the primary stores the signed client request without validating it, trusting that the forwarder backup has verified it.

I added the corresponding unit tests in `frontend_test.cpp`. 

Plus some minor changes:
- Moved some functions from `public:` to `private:` in `frontend.h`.
- Added curly braces under if/else statements in `frontend.h`.

### CODE REVIEW

File: tests/infra/rates.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 from loguru import logger as LOG
 
-COMMIT_COUNT_CUTOFF = 10
+COMMIT_COUNT_CUTOFF = 15
```
Comment (jumaffre) [inline tests/infra/rates.py]: I've increased this timeout a tiny bit as it seems that the forwarding performance test can sometimes take a little while to prepare the signed transactions (intermittent failures).

**Comment by cimetrics** [comment]
images

**Comment by cimetrics** [comment]
images

File: src/node/rpc/test/frontend_test.cpp [hunk 2]
```diff
@@ -506,16 +473,22 @@ TEST_CASE("No certs table")
...
+TEST_CASE("Signed read requests can be executed on backup")
 {
-  doctest::Context context;
-  context.applyCommandLine(argc, argv);
-  ::EverCrypt_AutoConfig2_init();
-  int res = context.run();
-  if (context.shouldExit())
-    return res;
-  return res;
+  prepare_callers();
```
Comment (olgavrou) [inline src/node/rpc/test/frontend_test.cpp]: Do we want to check here that the signature has not been stored in the kv?

File: src/node/rpc/frontend.h [hunk 3]
```diff
@@ -177,6 +183,64 @@ namespace ccf
       }
     }
 
+    void record_client_signature(
+      Store::Tx& tx, CallerId caller_id, SignedReq& signed_request)
+    {
+      if (request_storing_disabled)
+      {
+        signed_request.req.clear();
```
[the remark this answers is not in the document]
Comment (jumaffre) [reply src/node/rpc/frontend.h]: I believe the signed request storing is only disabled for our performance test (i.e. small bank) to make sure the amount of memory used is limited. The member voting is only used sparingly and should only affect the amount of memory used very little.

File: src/node/rpc/test/frontend_test.cpp (hunk 2 above)
Comment (jumaffre) [reply src/node/rpc/test/frontend_test.cpp]: Using the `BackupStubConsensus`, it is guaranteed that the transaction will not commit as `replicate()` will fail. If it did, we would get a `TX_FAILED_TO_REPLICATE` in the response.

File: src/node/rpc/frontend.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (olgavrou) [reply src/node/rpc/frontend.h]: But an application can choose to have signatures on and request storing off, in which case we might loose the member requests. The question I guess is if we want this to be a guarantee or a user option.

[the remark this answers is not in the document]
Comment (jumaffre) [reply src/node/rpc/frontend.h]: I don't think this is something that should be exposed to users. Signed votes and signed requests should always be recorded.

**Pull request merged**